### PR TITLE
Minor MicroBatch Class Cleanup and Return to Spark Structured Streaming Reader

### PR DESCRIPTION
**Pull request by kbendick**
I have been looking into adding support for a proper Spark Structured Streaming V2 Data Source

I can see that there was previously a PR, and there was a large amount of discussion around that PR. However, that PR goes all the way back to February 13th.

I would love to pick up where that work left off if @jerryshao is not working on it. I can see there was a very large amount of discussion and it appears as though some code that was suggested by @aokolnychyi and others was added in.

I started by looking through some of the current code and I noticed this (likely) doc typo. I also added in two tests for behavior checks as I was exploring the behavior of the `MicroBatches.from` builder.

I'd like to have one implementation that could eventually possibly be shared by Flink and continuous streaming, as the implementaton is size based and could be dynamic as well, but supporting Trigger.Once() and then triggering sized based batches with deletes in Spark should be prioritized imo as Flink is already working well. Though it should definitely be a source of inspiration when possible.

Please let me know what you think and I'll open several tickets for initial areas that I think need to be addressed based on the comments left on the other PR and we can have further discussion. cc @rdblue @aokolnychyi @HeartSaVioR @RussellSpitzer

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/MicroBatches.java [hunk 1]
```diff
@@ -145,7 +158,7 @@ public MicroBatch generate(int startFileIndex, long targetSizeInBytes, boolean i
     /**
      * Method to skip the manifest file in which the index is smaller than startFileIndex. For example, if the
      * index list is : (m1, 0), (m2, 3), (m3, 5), and startFileIndex is 4, then the returned manifest index list is:
-     * (m2, 3), (m3, 5).
+     * (m3, 5).
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/MicroBatches.java:161]: This update to this java doc comment is what inspired this PR.

File: core/src/test/java/org/apache/iceberg/TestMicroBatchBuilder.java [hunk 2]
```diff
@@ -140,13 +141,102 @@ public void testGenerateMicroBatchWithSmallTargetSize() {
...
-        .withPath(name + ".parquet")
-        .withFileSizeInBytes(10)
-        .withPartitionPath("data_bucket=0") // easy way to set partition data for now
-        .withRecordCount(1)
-        .build();
+            .withPath(name + ".parquet")
+            .withFileSizeInBytes(newFileSizeInBytes)
+            .withPartitionPath("data_bucket=0") // easy way to set partition data for now
+            .withRecordCount(1)
+            .build();
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestMicroBatchBuilder.java:234]: Looks like you indentation settings are off. This should be 2 indents (4 spaces) from the start of `return`.

Comment (kbendick) [reply core/src/test/java/org/apache/iceberg/TestMicroBatchBuilder.java]: Ok this has been fixed. Looks correct now based on the diff!

File: core/src/main/java/org/apache/iceberg/MicroBatches.java (hunk 1 above)
Comment (jerryshao) [reply core/src/main/java/org/apache/iceberg/MicroBatches.java:161]: The change is not correct. If the `startFileIndex` is 4, then "(m2, 3)" should not be skipped, the 4th index is included in "(m2, 3)"

File: core/src/test/java/org/apache/iceberg/TestMicroBatchBuilder.java [hunk 3]
```diff
@@ -140,15 +141,104 @@ public void testGenerateMicroBatchWithSmallTargetSize() {
...
+    Assert.assertEquals(4, batch0.endFileIndex());
+    Assert.assertEquals(40, batch0.sizeInBytes());
+    filesMatch(Lists.newArrayList("A", "B", "C", "D"), filesToScan(batch0.tasks()));
+    Assert.assertFalse(batch0.lastIndexOfSnapshot());
+
+    // Concurrent write sometime after the start of the last batch and before the next batch.
+    final long sizeOfFileF = 25L;
+    add(table.newAppend(),
+            Collections.singletonList(fileWithSize("F", sizeOfFileF)));
+    Assert.assertEquals(2L, table.currentSnapshot().snapshotId());
```
Comment (jerryshao) [inline core/src/test/java/org/apache/iceberg/TestMicroBatchBuilder.java:203]: This will generate a new snapshot, seems unrelated to the `MicroBatch` processed before, not sure what's the purpose of this testing.

**Comment by kbendick** [comment]
I'm closing this PR as my understanding is that @jerryshao was going to continue to work on this (and the work might have been finished by now). This is to make it easier for the release managers to know which PRs need to be included or that the release needs to be blocked on waiting for this (which it does not).
